### PR TITLE
Allow getattr to work in objmode output type spec

### PR DESCRIPTION
**Pull request by sklam**
Related to https://github.com/numba/numba/issues/6668

### CODE REVIEW

File: numba/tests/test_withlifting.py [hunk 1]
```diff
@@ -825,6 +825,37 @@ def global_var():
...
+    def test_objmode_gv_mod_attr_error(self):
+        @njit
+        def moderror():
+            with objmode(val=types.THIS_DOES_NOT_EXIST):
+                val = 12.3
+            return val
+        with self.assertRaises(errors.CompilerError) as raises:
+            moderror()
+        self.assertIn("Getattr cannot be resolved at compile-time",
+                      str(raises.exception))
```
Comment (gmarkall) [inline numba/tests/test_withlifting.py:857]: Is it worth checking that the reported getattr that can't be resolved is as expected, e.g.:

```suggestion
                      str(raises.exception))
        self.assertIn("types.THIS_DOES_NOT_EXIST):\n            ^",
                      str(raises.exception))
```

**Review by gmarkall — COMMENTED** [review]
The changes look good to me - there is one small suggestion for the error message check on the diff.

I presume this relates only to use case 1 in #6668?

File: numba/tests/test_withlifting.py (hunk 1 above)
Comment (sklam) [reply numba/tests/test_withlifting.py:857]: The suggested check depends on the error reporting options; e.g. i'm seeing colorama added codes and the suggested check is failing. I don't think it's worth writing all the regex just to match it correctly given that the current error message can only be coming from the code I added.

**Comment by sklam** [comment]
@gmarkall, thx for reviewing. And, yes, this is the use-case 1 for #6668.

File: numba/tests/test_withlifting.py (hunk 1 above)
Comment (gmarkall) [reply numba/tests/test_withlifting.py:857]: OK, makes sense.

**Review by gmarkall — APPROVED** [review]
Looks good to me.

**Comment by gmarkall** [comment]
Not sure if @esc @stuartarchibald also want to look at this?

**Comment by gmarkall** [comment]
Hit the wrong button, sorry.

File: numba/tests/test_withlifting.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/tests/test_withlifting.py:857]: I bumped into this whilst testing:
```python
from numba import njit, objmode
import numpy as np

@njit
def moderror(pred):
    if pred:
        x = 'intp'
    else:
        x = 'float32'
    with objmode(val=x):
        val = 12.3
    return val

moderror(False)
```
not necessarily to be fixed here, but https://github.com/numba/numba/blob/d89d34105cc0e1dc9402b14bbc0e4a7abf3c0a3e/numba/core/transforms.py#L382 should guard for multiply defined variables.

File: numba/core/withcontexts.py [hunk 2]
```diff
@@ -220,12 +222,22 @@ def _legalize_args(self, args, kwargs, loc, func_globals, func_closures):
                         f"Global {v.name!r} is not defined"
                     )
                 typeanns[k] = v
+            elif isinstance(v, ir.Expr) and v.op == "getattr":
+                try:
+                    base_obj = func_ir.infer_constant(v.value)
+                    typ = getattr(base_obj, v.attr)
+                except (errors.ConstantInferenceError, AttributeError):
+                    raise errors.CompilerError(
+                        f"Getattr cannot be resolved at compile-time",
```
Comment (stuartarchibald) [inline numba/core/withcontexts.py:231]: Should this include information from `base_obj` and `v.value`? From e.g. this:
```python
from numba import njit, objmode, types
import numpy as np

@njit
def moderror():
    with objmode(v1=types.NOT_A_TYPE, v2=types.intp, v3=types.float32):
        v1 = 12.3
        v2 = 12.3
        v3 = 12.3
    return v1, v2, v3

moderror()
```
gives:
```
numba.core.errors.CompilerError: Failed in nopython mode pipeline (step: Handle with contexts)
Getattr cannot be resolved at compile-time

File "test_6680.py", line 6:
def moderror():
    with objmode(v1=types.NOT_A_TYPE, v2=types.intp, v3=types.float32):
    ^
```
whilst entirely obvious in this contrived case, in general would it be clear which of the `v{1,2,3}` is the issue?

Comment (sklam) [reply numba/core/withcontexts.py]: I've made the msg better. The above will now show:

```
CompilerError: Failed in nopython mode pipeline (step: Handle with contexts)
Error handling objmode argument 'v1'. Getattr cannot be resolved at compile-time.

File "<ipython-input-1-a794355d32dc>", line 6:
def moderror():
    with objmode(v1=types.NOT_A_TYPE, v2=types.intp, v3=types.float32):
    ^
```

Comment (stuartarchibald) [reply numba/core/withcontexts.py]: Thanks for changing that, the latest patch makes this case much clearer.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes, looks good.

**Pull request merged**